src/java.desktop/share/classes/sun/font/HBShaper.java

```diff
@@ -0,0 +1,617 @@
+    ) {
+
+        FontStrike strike = scopedVars.get().fontStrike();
+        Point2D.Float pt = strike.getGlyphMetrics(glyph);
+        return (pt != null) ? HBFloatToFixed(pt.x) : 0;
+    }
+
+    private static int get_glyph_v_advance(
+        MemorySegment font_ptr,   /* Not used */
+        MemorySegment font_data,  /* Not used */
+        int glyph,
+        MemorySegment user_data  /* Not used */
+    ) {
+
+        FontStrike strike = scopedVars.get().fontStrike();
+        Point2D.Float pt = strike.getGlyphMetrics(glyph);
+        return (pt != null) ? HBFloatToFixed(pt.y) : 0;
+    }
+
+    /*
+     * This class exists to make the code that uses it less verbose
+     */
+    private static class IntPtr {
+        MemorySegment seg;
+        IntPtr(MemorySegment seg) {
+        }
+
+        void set(int i) {
+            seg.setAtIndex(JAVA_INT, 0, i);
+        }
+    }
+
+    private static int get_glyph_contour_point(
+        MemorySegment font_ptr,   /* Not used */
+        MemorySegment font_data,  /* Not used */
+        int glyph,
+        int point_index,
+        MemorySegment x_ptr,     /* ptr to return x */
+        MemorySegment y_ptr,     /* ptr to return y */
+        MemorySegment user_data  /* Not used */
+    ) {
+        IntPtr x = new IntPtr(x_ptr);
+        IntPtr y = new IntPtr(y_ptr);
+
+        if ((glyph & 0xfffe) == 0xfffe) {
+            x.set(0);
+            y.set(0);
+            return 1;
+        }
+
+        FontStrike strike = scopedVars.get().fontStrike();
+        Point2D.Float pt = ((PhysicalStrike)strike).getGlyphPoint(glyph, point_index);
+        x.set(HBFloatToFixed(pt.x));
+        y.set(HBFloatToFixed(pt.y));
+
+       return 1;
+    }
+
+    record ScopedVars (
+        Font2D font,
+        FontStrike fontStrike,
+        GVData gvData,
+        Point2D.Float point) {}
+
+    static final ScopedValue<ScopedVars> scopedVars = ScopedValue.newInstance();
+
+    static void shape(
+        Font2D font2D,
+        FontStrike fontStrike,
+        float ptSize,
+        float[] mat,
+        MemorySegment hbface,
+        char[] text,
+        GVData gvData,
+        int script,
+        int offset,
+        int limit,
+        int baseIndex,
+        Point2D.Float startPt,
+        int flags,
+        int slot) {
+
+        /*
+         * ScopedValue is needed so that call backs into Java during
+         * shaping can locate the correct instances of these to query or update.
+         * The alternative of creating bound method handles is far too slow.
+         */ 
+        ScopedVars vars = new ScopedVars(font2D, fontStrike, gvData, startPt);
+        ScopedValue.where(scopedVars, vars)
+                   .run(() -> {
+
+            try (Arena arena = Arena.ofConfined()) {
+
+                float startX = (float)startPt.getX();
+                float startY = (float)startPt.getY();
+
+                MemorySegment matrix = arena.allocateArray(JAVA_FLOAT, mat.length);
```

Comment:
There should be an overload of `allocateArray` which takes a Java array directly and then copies it off-heap after allocation. In Java 22 this method is called `allocateFrom` and is much more optimized (as it avoids zeroing of memory).